autograder/handgrading/tests/test_views/test_handgrading_result_views.py

```diff
@@ -670,47 +670,65 @@ def test_get_handgrading_results_exclude_staff(self):
 
     def test_handgrader_get_handgrading_results_default_page_size(self):
-        self.do_handgrading_results_test(self.handgrader, num_results=2)
+        self.do_handgrading_results_test(self.handgrader, num_groups=2)
 
     def test_staff_get_paginated_handgrading_results_custom_page_size(self):
-        self.do_handgrading_results_test(self.staff, num_results=3, page_size=1)
+        self.do_handgrading_results_test(self.staff, num_groups=3, page_size=1)
 
     def test_handgrader_get_paginated_handgrading_results_custom_page_size(self):
-        self.do_handgrading_results_test(self.handgrader, num_results=3, page_size=1)
+        self.do_handgrading_results_test(self.handgrader, num_groups=3, page_size=1)
 
     def test_staff_get_paginated_handgrading_results_specific_page(self):
-        self.do_handgrading_results_test(self.staff, num_results=4, page_size=3, page_num=2)
+        self.do_handgrading_results_test(self.staff, num_groups=4, page_size=3, page_num=2)
 
     def test_handgrader_get_paginated_handgrading_results_specific_page(self):
-        self.do_handgrading_results_test(self.handgrader, num_results=4, page_size=3, page_num=2)
+        self.do_handgrading_results_test(self.handgrader,
+                                         num_groups=4, page_size=3, page_num=2)
 
     def test_get_paginated_results_some_groups_have_no_result(self):
-        self.do_handgrading_results_test(self.handgrader, num_results=2, num_groups=4)
+        self.do_handgrading_results_test(
+            self.handgrader, num_groups=4, submission_statuses=[
+                Submission.GradingStatus.finished_grading] * 2 + [None] * 2)
 
     def test_non_staff_non_handgrader_get_handgrading_results_permission_denied(self):
         [student] = obj_build.make_student_users(self.course, 1)
         self.client.force_authenticate(student)
         response = self.client.get(reverse('handgrading_results', kwargs={'pk': self.project.pk}))
         self.assertEqual(status.HTTP_403_FORBIDDEN, response.status_code)
 
-    def do_handgrading_results_test(self, user: User, *,
-                                    num_results: int, num_groups: int = None,
+    def test_handgrader_get_handgrading_results_only_rejected_submissions(self):
+        self.do_handgrading_results_test(self.handgrader, num_groups=3, submission_statuses=[
+            Submission.GradingStatus.finished_grading] + 2 * [Submission.GradingStatus.rejected])
+
+    def test_staff_get_handgrading_results_only_rejected_submissions(self):
+        self.do_handgrading_results_test(self.staff, num_groups=3, submission_statuses=[
+            Submission.GradingStatus.finished_grading] + [Submission.GradingStatus.rejected] * 2)
+
+    def do_handgrading_results_test(self, user: User, *, num_groups: int = None,
+                                    submission_statuses:
+                                    list[ag_models.Submission.GradingStatus | None] | None = None,
```

Comment:
I like the direction this is going. Rather than an overloaded use of None (where it can tell us something about the submission or the handgrading result), maybe we should come up with a class to encode this? Maybe something like:

```python
@dataclass
class _GroupHGDataConfig:
    submission_status: ag_models.Submission.GradingStatus = ag_models.Submission.GradingStatus.finished_grading
    has_handgrading_result: bool = True

    def __post_init__(self):
        if has_handgrading_result:
            assert self.submission_status == ag_models.Submission.GradingStatus.finished_grading, 'Submissions with a handgrading result must have finished_grading status'

        if self.submission_status != ag_models.Submission.GradingStatus.finished_grading:
            assert not self.has_handgrading_result, 'Submissions with a status other than "finished_grading" must not have a handgrading result'
```

Then, I think the cases we have to consider are slightly different from the current change. We need to be able to handle:
- A finished_grading submission with a handgrading result
- A finished_grading submission without a handgrading result (has_handgradeable_submission would still be *true* in this case)
- A submission with any status other than finished_grading (and consequently no handgrading result either). has_handgradeable_submission would be false in this case. Note that we don't need to special case rejected for this: you can use `obj_buid.make_submission(group=group, status=status)` to create a submission with the proper status.

After that, we can add copies of the `test_handgrader/staff_get_handgrading_results_only_rejected_submissions` tests for `error` and `removed_from_queue` statuses.

P.S. Sorry I didn't see your question about the type of `submission_statuses` sooner. For some reason I didn't get an email notification about it. In the future, if I don't get back to you within 2 business days, feel free to email me.

Reply:
Thanks for your comment, @james-perretta 

I presume we are using `_GroupHGDataConfig` to indicate what kind of submission statuses we are testing and the corresponding `has_handgrading_result`

I am just not sure how we may differentiate these two cases
> * A finished_grading submission with a handgrading result
> * A finished_grading submission without a handgrading result (has_handgradeable_submission would still be true in this case)

by using `_GroupHGDataConfig` alone. 

I presume in the second case, a second group would `make_finished_submission` but the expected `data['handgrading_result']` would be `None` while `data['has_handgradeable_submission']` would still be true. I don't think we currently have this test case, and I add that once I devise some ways to differentiate between `finished_grading` with or without `handgrading_result`.